kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+                    Error::change_data_feed_incompatible_schema(&self.schema, &schema)
+                );
+
+                let table_properties = TableProperties::from(configuration);
+                require!(
+                    table_properties.enable_change_data_feed.unwrap_or(false),
+                    Error::change_data_feed_unsupported(self.commit_file.version)
+                )
+            }
+        }
+
+        // We resolve the remove deletion vector map after visiting the entire commit.
+        if self.has_cdc_action {
+            self.remove_dvs.clear();
+        } else {
+            // The only (path, deletion_vector) pairs we must track are ones whose path is the
+            // same as an `add` action.
+            self.remove_dvs
+                .retain(|rm_path, _| add_paths.contains(rm_path));
+        }
+        Ok(())
+    }
+
+    fn into_scan_batches(
+        self,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+        let Self {
+            has_cdc_action,
+            remove_dvs,
+            commit_file,
+            json_handler,
+            timestamp: _,
+            filter,
+            expression_handler,
+            schema: _,
+        } = self;
+        let remove_dvs = Arc::new(remove_dvs);
+
+        let schema = FileActionSelectionVisitor::schema()?;
+        let action_iter =
+            json_handler.read_json_files(&[commit_file.location.clone()], schema, None)?;
+
+        let result = action_iter.map(move |actions| -> DeltaResult<_> {
+            let actions = actions?;
+
+            // Apply data skipping to get back a selection vector for actions that passed skipping.
+            // We start our selection vector based on what was filtered. We will add to this vector
+            // below if a file has been removed. Note: None implies all files passed data skipping.
+            let selection_vector = match &filter {
+                Some(filter) => filter.apply(actions.as_ref())?,
+                None => vec![true; actions.len()],
+            };
+
+            let mut visitor =
+                FileActionSelectionVisitor::new(&remove_dvs, selection_vector, has_cdc_action);
+            visitor.visit_rows_of(actions.as_ref())?;
+
+            let data = expression_handler
+                .get_evaluator(
+                    get_log_add_schema().clone(),
+                    get_add_transform_expr(),
+                    scan_row_schema().into(),
+                )
+                .evaluate(actions.as_ref())?;
+            Ok(TableChangesScanData {
+                data,
+                selection_vector: visitor.selection_vector,
+                remove_dv: Some(remove_dvs.clone()),
+            })
+        });
+        Ok(result)
+    }
+}
+
+// This is a visitor used in the prepare phase of [`LogReplayScanner`]. See
+// [`LogReplayScanner::prepare_phase`] for details usage.
+struct PreparePhaseVisitor<'a> {
+    scanner: &'a mut LogReplayScanner,
```

Comment:
Shouild I reconsider having the scanner as a parameter? I wonder if it makes it harder to test this visitor?

Reply:
See other comment -- I don't think a scanner even needs to exist yet (so we should indeed not have it as a param)

Reply:
Cool. I've got mutable references to the relevant fields just cuz that's the fewest LOC.